### ISSUE TITLE
Enhancement request: include IDs & criteriaReference on population & episode results output

### ISSUE THREAD

**Issue by nmorasb** [projecttacoma/fqm-execution#146]
Would it be possible to include IDs and criteriaReference fields in the execution output for populationResults, episodeResults, and stratificationResults? 

Currently it is difficult to match up observation and stratification output to the inputs for comparison of expected vs actual values. 

Outputs do now include the criteria expression, which helps, but it is possible for users to use the same expression multiple times (for observations). 

Similarly, they could have multiple stratifications on the same population. The outputs are then difficult to map back to the inputs for comparison of expected vs actual values.

**Comment by mgramigna** [comment]
Hi @nmorasb, I agree this would be a good enhancement.

Just to confirm, are you referring to the ID of the population as defined in the measure?, e.g.

```
{
  // ...
  "population": [
    {
      "id": "some-unique-id",
      "code": {
        "coding": [
          {
            "system": "http://terminology.hl7.org/CodeSystem/measure-population",
            "code": "initial-population",
            "display": "Initial Population"
          }
        ]
      },
      "criteria": {
        "language": "text/cql.identifier",
        "expression": "ipp"
      }
    }
    // ...
  ]
}
```

Should result in the following for population results?

```
{
  // ...
  "populationResults": [
    {
      "id": "some-unique-id",
      "populationType": "initial-population",
      "criteriaExpression": "ipp",
      "result": true
    }
    // ...
  ]
}
```

**Comment by mgramigna** [comment]
I've implemented this in #161. Here is a snippet of the new output (taken from running a ratio encounter-based measure from #159):

Does this help the ask here?

```
{
  "groupId": "group-ratio-encounter",
  "populationResults": [
    {
      "populationType": "initial-population",
      "criteriaExpression": "ipp",
      "result": true,
      "populationId": "42875cf3-c47a-4d88-ab83-43b2f7039751"
    },
    {
      "populationType": "denominator",
      "criteriaExpression": "denom",
      "result": true,
      "populationId": "3f48596a-85d2-4c5c-b0d3-ec2698d1872f"
    },
    {
      "populationType": "denominator-exclusion",
      "result": false,
      "populationId": "3515943e-a88c-4fe8-8398-25c9cb939273"
    },
    {
      "populationType": "numerator",
      "criteriaExpression": "num",
      "result": true,
      "populationId": "662c463c-695c-43ba-a9f5-a0b4d650d460"
    },
    {
      "populationType": "numerator-exclusion",
      "result": false,
      "populationId": "21c2f52d-a08a-4392-b905-4b51f242c1fe"
    },
    {
      "populationType": "measure-observation",
      "criteriaExpression": "fun2",
      "result": true,
      "populationId": "917c1c18-e36c-48e4-b2b3-989d49bdb8cc",
      "criteriaReferenceId": "3f48596a-85d2-4c5c-b0d3-ec2698d1872f"
    },
    {
      "populationType": "measure-observation",
      "criteriaExpression": "fun3",
      "result": true,
      "populationId": "5c6f858e-3f31-4fc3-8814-a2b011039931",
      "criteriaReferenceId": "662c463c-695c-43ba-a9f5-a0b4d650d460"
    }
  ],
  "episodeResults": [
    {
      "episodeId": "Encounter2",
      "populationResults": [
        {
          "populationType": "initial-population",
          "criteriaExpression": "ipp",
          "result": true,
          "populationId": "42875cf3-c47a-4d88-ab83-43b2f7039751"
        },
        {
          "populationType": "denominator",
          "criteriaExpression": "denom",
          "result": true,
          "populationId": "3f48596a-85d2-4c5c-b0d3-ec2698d1872f"
        },
        {
          "populationType": "denominator-exclusion",
          "result": false,
          "populationId": "3515943e-a88c-4fe8-8398-25c9cb939273"
        },
        {
          "populationType": "numerator",
          "criteriaExpression": "num",
          "result": false,
          "populationId": "662c463c-695c-43ba-a9f5-a0b4d650d460"
        },
        {
          "populationType": "numerator-exclusion",
          "result": false,
          "populationId": "21c2f52d-a08a-4392-b905-4b51f242c1fe"
        },
        {
          "populationType": "measure-observation",
          "criteriaExpression": "fun2",
          "result": true,
          "populationId": "917c1c18-e36c-48e4-b2b3-989d49bdb8cc",
          "criteriaReferenceId": "3f48596a-85d2-4c5c-b0d3-ec2698d1872f",
          "observations": [2]
        },
        {
          "populationType": "measure-observation",
          "criteriaExpression": "fun3",
          "result": false,
          "populationId": "5c6f858e-3f31-4fc3-8814-a2b011039931",
          "criteriaReferenceId": "662c463c-695c-43ba-a9f5-a0b4d650d460"
        }
      ]
    },
    {
      "episodeId": "Encounter3",
      "populationResults": [
        {
          "populationType": "initial-population",
          "criteriaExpression": "ipp",
          "result": true,
          "populationId": "42875cf3-c47a-4d88-ab83-43b2f7039751"
        },
        {
          "populationType": "denominator",
          "criteriaExpression": "denom",
          "result": true,
          "populationId": "3f48596a-85d2-4c5c-b0d3-ec2698d1872f"
        },
        {
          "populationType": "denominator-exclusion",
          "result": false,
          "populationId": "3515943e-a88c-4fe8-8398-25c9cb939273"
        },
        {
          "populationType": "numerator",
          "criteriaExpression": "num",
          "result": true,
          "populationId": "662c463c-695c-43ba-a9f5-a0b4d650d460"
        },
        {
          "populationType": "numerator-exclusion",
          "result": false,
          "populationId": "21c2f52d-a08a-4392-b905-4b51f242c1fe"
        },
        {
          "populationType": "measure-observation",
          "criteriaExpression": "fun2",
          "result": true,
          "populationId": "917c1c18-e36c-48e4-b2b3-989d49bdb8cc",
          "criteriaReferenceId": "3f48596a-85d2-4c5c-b0d3-ec2698d1872f",
          "observations": [2]
        },
        {
          "populationType": "measure-observation",
          "criteriaExpression": "fun3",
          "result": true,
          "populationId": "5c6f858e-3f31-4fc3-8814-a2b011039931",
          "criteriaReferenceId": "662c463c-695c-43ba-a9f5-a0b4d650d460",
          "observations": [4]
        }
      ]
    }
  ]
}
```

**Comment by mgramigna** [comment]
And here's an example from a measure that had stratifiers:

```
{
  "groupId": "cfae4813-f901-4abb-8088-2a4efe3518c3",
  "populationResults": [
    {
      "populationType": "initial-population",
      "criteriaExpression": "ipp",
      "result": true,
      "populationId": "0bde2928-8ca5-4b50-b73d-8a33107c5ee7"
    },
    {
      "populationType": "denominator",
      "criteriaExpression": "denom",
      "result": true,
      "populationId": "c2088e04-44a4-4b60-b0f4-15d636777fa1"
    },
    {
      "populationType": "numerator",
      "criteriaExpression": "numer",
      "result": true,
      "populationId": "c6253169-503f-4a08-be5f-ddd0397f6669"
    }
  ],
  "episodeResults": [
    {
      "episodeId": "Encounter2",
      "populationResults": [
        {
          "populationType": "initial-population",
          "criteriaExpression": "ipp",
          "result": true,
          "populationId": "0bde2928-8ca5-4b50-b73d-8a33107c5ee7"
        },
        {
          "populationType": "denominator",
          "criteriaExpression": "denom",
          "result": true,
          "populationId": "c2088e04-44a4-4b60-b0f4-15d636777fa1"
        },
        {
          "populationType": "numerator",
          "criteriaExpression": "numer",
          "result": true,
          "populationId": "c6253169-503f-4a08-be5f-ddd0397f6669"
        }
      ],
      "stratifierResults": [
        {
          "strataCode": "strata-1",
          "result": false
        },
        {
          "strataCode": "strata-2",
          "result": true,
          "strataId": "strat2-unique-id"
        }
      ]
    },
    {
      "episodeId": "Encounter3",
      "populationResults": [
        {
          "populationType": "initial-population",
          "criteriaExpression": "ipp",
          "result": true,
          "populationId": "0bde2928-8ca5-4b50-b73d-8a33107c5ee7"
        },
        {
          "populationType": "denominator",
          "criteriaExpression": "denom",
          "result": true,
          "populationId": "c2088e04-44a4-4b60-b0f4-15d636777fa1"
        },
        {
          "populationType": "numerator",
          "criteriaExpression": "numer",
          "result": true,
          "populationId": "c6253169-503f-4a08-be5f-ddd0397f6669"
        }
      ],
      "stratifierResults": [
        {
          "strataCode": "strata-1",
          "result": true,
          "strataId": "strat1-unique-id"
        },
        {
          "strataCode": "strata-2",
          "result": false
        }
      ]
    }
  ],
  "stratifierResults": [
    {
      "result": true,
      "strataCode": "strata-1"
    },
    {
      "result": true,
      "strataCode": "strata-2",
      "strataId": "strat2-unique-id"
    }
  ]
}
```
